// otplib/hotp.js
/**
 * otplib-hotp
 *
 * @author Gerald Yeo <contact@fusedthought.com>
 * @version: 9.0.0
 * @license: MIT
 **/
'use strict';

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var otplibCore = require('./core');

class HOTP {
  constructor() {
    this._options = this.defaultOptions;
  }
  get defaultOptions() {
    return {};
  }
  set options(opt = {}) {
    if (opt) {
      this._options = _extends({}, this._options, opt);
    }
  }
  get options() {
    return _extends({}, this._options);
  }
  get optionsAll() {
    return otplibCore.hotpOptions(this._options);
  }
  resetOptions() {
    this._options = this.defaultOptions;
    return this;
  }
  generate(secret, counter) {
    const opt = this.optionsAll;
    return otplibCore.hotpToken(secret || opt.secret, counter, opt);
  }
  check(token, secret, counter) {
    const opt = this.optionsAll;
    return otplibCore.hotpCheck(token, secret || opt.secret, counter, opt);
  }
  verify(opts) {
    if (typeof opts !== 'object' || opts == null) {
      return false;
    }
    return this.check(opts.token, opts.secret, opts.counter);
  }
}
HOTP.prototype.HOTP = HOTP;

var index = new HOTP();

module.exports = index;
